File: cmd/model/markov-chain.go
package model

import (
	"math/rand"
	"strings"
	"sync"
)

type MarkovChain struct {
	ID             string
	ReplyRate      int
	Pings          bool
	State          map[string]map[string]int
	MessageCounter int
	MediaStorage   *MediaStorage
	mu             sync.RWMutex
}

func NewMarkovChain(id string, replyRate int, pings bool, messages []string) *MarkovChain {
	mc := &MarkovChain{
		ID:           id,
		ReplyRate:    replyRate,
		Pings:        pings,
		State:        make(map[string]map[string]int),
		MediaStorage: NewMediaStorage(id, nil, nil, nil),
	}
	mc.ProvideData(messages)
	return mc
}

func (mc *MarkovChain) ProvideData(messages []string) {
	for _, message := range messages {
		mc.UpdateState(message)
	}
}

func (mc *MarkovChain) UpdateState(message string) {
	if strings.HasPrefix(message, "https://") {
		mc.MediaStorage.AddMedia(message)
		return
	}

	mc.MessageCounter++
	tokens := mc.Tokenize(message)
	for i := 0; i < len(tokens)-1; i++ {
		currentWord := tokens[i]
		nextWord := tokens[i+1]

		mc.mu.Lock()
		if mc.State[currentWord] == nil {
			mc.State[currentWord] = make(map[string]int)
		}
		mc.State[currentWord][nextWord]++
		mc.mu.Unlock()
	}
}

func (mc *MarkovChain) GenerateText(startWord string, length int) string {
	mc.mu.RLock()
	defer mc.mu.RUnlock()

	currentWord := startWord
	var generatedText strings.Builder
	generatedText.WriteString(currentWord)

	for i := 0; i < length; i++ {
		nextWords, exists := mc.State[currentWord]
		if !exists {
			break
		}

		var nextWordArray []string
		var nextWordWeights []int
		for word, weight := range nextWords {
			nextWordArray = append(nextWordArray, word)
			nextWordWeights = append(nextWordWeights, weight)
		}

		smoothedWeights := make([]float64, len(nextWordWeights))
		for i, weight := range nextWordWeights {
			smoothedWeights[i] = float64(weight+1) / float64(mc.MessageCounter+len(nextWordArray))
		}

		nextWord := mc.StochasticChoice(nextWordArray, smoothedWeights)
		currentWord = nextWord
		generatedText.WriteString(" " + currentWord)
	}

	if !mc.Pings {
		// Replace @mentions, add your regex logic here
		generatedTextStr := generatedText.String()
		return strings.ReplaceAll(generatedTextStr, "<@&?\\w+>", "$1")
	}
	return generatedText.String()
}

func (mc *MarkovChain) Delete(message string) {
	if strings.HasPrefix(message, "https://") {
		mc.MediaStorage.RemoveMedia(message)
		return
	}

	tokens := mc.Tokenize(message)
	for i := 0; i < len(tokens)-1; i++ {
		currentWord := tokens[i]
		nextWord := tokens[i+1]

		if nextWordMap, exists := mc.State[currentWord]; exists {
			if _, exists := nextWordMap[nextWord]; exists {
				delete(nextWordMap, nextWord)
				// Clean up the map if it's empty
				if len(nextWordMap) == 0 {
					delete(mc.State, currentWord)
				}
			}
		}
	}
}

func (mc *MarkovChain) StochasticChoice(options []string, weights []float64) string {
	totalWeight := 0.0
	for _, weight := range weights {
		totalWeight += weight
	}
	randomWeight := rand.Float64() * totalWeight
	var weightSum float64
	for i, option := range options {
		weightSum += weights[i]
		if randomWeight <= weightSum {
			return option
		}
	}
	return options[len(options)-1]
}

func (mc *MarkovChain) Tokenize(text string) []string {
	tokens := strings.Fields(text)
	var filteredTokens []string
	for _, token := range tokens {
		if len(token) > 0 {
			filteredTokens = append(filteredTokens, token)
		}
	}
	return filteredTokens
}

func (mc *MarkovChain) Talk(length int) string {
	mc.mu.RLock()
	defer mc.mu.RUnlock()

	keys := make([]string, 0, len(mc.State))
	for key := range mc.State {
		keys = append(keys, key)
	}

	randomIndex := rand.Intn(len(keys))
	startingWord := keys[randomIndex]

	return mc.GenerateText(startingWord, length)
}

File: src/domain/model/markov.chain.ts
import { MediaStorage } from './media.storage';

export class MarkovChain {
	public mediaStorage: MediaStorage;
	state: MarkovState;
	messageCounter = 0;

	constructor(
		public id: string,
		public replyRate = 10,
		public pings = true,
		messages: string[] = []
	) {
		this.mediaStorage = new MediaStorage(this.id);
		this.state = {};
		this.provideData(messages);
	}

	provideData(messages: string[]): void {
		messages.forEach((message) => this.updateState(message));
	}

	updateState(message: string): void {
		if (message.startsWith('https://')) {
			this.mediaStorage.addMedia(message);
			return;
		}
		this.messageCounter++;
		const tokens = this.tokenize(message);
		for (let i = 0; i < tokens.length - 1; i++) {
			const currentWord = tokens[i];
			const nextWord = tokens[i + 1];
			this.state[currentWord] ??= {};
			this.state[currentWord][nextWord] ??= 0;
			this.state[currentWord][nextWord]++;
		}
	}

	generateText(startWord: string, length: number): string {
		let currentWord = startWord;
		let generatedText = currentWord;
		for (let i = 0; i < length; i++) {
			const nextWords = this.state[currentWord];
			if (!nextWords) {
				break;
			}
			const nextWordArray = Object.keys(nextWords);
			const nextWordWeights = Object.values(nextWords);
			// ? Laplace smoothing
			const smoothedWeights = nextWordWeights.map(
				(weight) => (weight + 1) / (this.messageCounter + nextWordArray.length)
			);
			currentWord = this.stochasticChoice(nextWordArray, smoothedWeights);
			generatedText += ` ${currentWord}`;
		}
		if (!this.pings) {
			generatedText = generatedText.replace(/<(@&?\w+)>/g, '$1');
		}
		return generatedText;
	}

	talk(length: number): string {
		const keys = Object.keys(this.state);
		const randomIndex = Math.floor(Math.random() * keys.length);
		const starterWord = keys[randomIndex];
		return this.generateText(starterWord, length).trim();
	}

	delete(message: string) {
		if (message.startsWith('https://')) {
			this.mediaStorage.removeMedia(message);
		}
		const tokens = this.tokenize(message);
		for (let i = 0; i < tokens.length - 1; i++) {
			const currentWord = tokens[i];
			const nextWord = tokens[i + 1];
			if (this.state[currentWord]?.[nextWord]) {
				delete this.state[currentWord][nextWord];
			}
		}
	}

	private tokenize(text: string) {
		const tokens = text.split(/\s+/);
		return tokens.filter((token) => token.length > 0);
	}

	private stochasticChoice(options: string[], weights: number[]): string {
		const totalWeight = weights.reduce((a, b) => a + b, 0);
		const randomWeight = Math.random() * totalWeight;
		let weightSum = 0;
		for (let i = 0; i < options.length; i++) {
			weightSum += weights[i];
			if (randomWeight <= weightSum) {
				return options[i];
			}
		}
		return options[options.length - 1];
	}
}
